runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/source/FlinkSourceReaderBase.java

```diff
@@ -0,0 +1,391 @@
+    this.waitingForSplitChangeFuture = new CompletableFuture<>();
+    this.idleTimeoutCountingDown = false;
+    // TODO: Remove the casting and use SourceReaderMetricGroup after minimum FLink version is
+    // upgraded to 1.14 and above.
+    this.numRecordsInCounter = FlinkSourceCompat.getNumRecordsInCounter(context);
+  }
+
+  @Override
+  public void start() {}
+
+  @Override
+  public List<FlinkSourceSplit<T>> snapshotState(long checkpointId) {
+    checkExceptionAndMaybeThrow();
+    // Add all the source splits whose readers haven't been created.
+    List<FlinkSourceSplit<T>> splitsState = new ArrayList<>(sourceSplits);
+
+    // Add all the source splits being actively read.
+    beamSourceReaders.forEach(
+        (splitId, readerAndOutput) -> {
+          Source.Reader<T> reader = readerAndOutput.reader;
+          if (reader instanceof BoundedSource.BoundedReader) {
+            // Sometimes users may decide to run a bounded source in streaming mode as "finite
+            // stream."
+            // For bounded source, the checkpoint granularity is the entire source split.
+            // So, in case of failure, all the data from this split will be consumed again.
+            splitsState.add(new FlinkSourceSplit<>(splitId, reader.getCurrentSource()));
+          } else if (reader instanceof UnboundedSource.UnboundedReader) {
+            // The checkpoint for unbounded sources is fine granular.
+            byte[] checkpointState =
+                getAndEncodeCheckpointMark((UnboundedSource.UnboundedReader<OutputT>) reader);
+            splitsState.add(
+                new FlinkSourceSplit<>(splitId, reader.getCurrentSource(), checkpointState));
+          }
+        });
+    return splitsState;
+  }
+
+  @Override
+  public CompletableFuture<Void> isAvailable() {
+    checkExceptionAndMaybeThrow();
+    if (!sourceSplits.isEmpty() || !beamSourceReaders.isEmpty()) {
+      // There are still live readers.
+      return isAvailableForAliveReaders();
+    } else if (noMoreSplits) {
+      // All the splits have been read, wait for idle timeout.
+      checkIdleTimeoutAndMaybeStartCountdown();
+      return idleTimeoutFuture;
+    } else {
+      // There is no live readers, waiting for new split assignments or no more splits notification.
+      if (waitingForSplitChangeFuture.isDone()) {
+        waitingForSplitChangeFuture = new CompletableFuture<>();
+      }
+      return waitingForSplitChangeFuture;
+    }
+  }
+
+  @Override
+  public void notifyNoMoreSplits() {
+    checkExceptionAndMaybeThrow();
+    LOG.info("Received NoMoreSplits signal from enumerator.");
+    noMoreSplits = true;
+    waitingForSplitChangeFuture.complete(null);
+  }
+
+  @Override
+  public void addSplits(List<FlinkSourceSplit<T>> splits) {
+    checkExceptionAndMaybeThrow();
+    LOG.info("Adding splits {}", splits);
+    sourceSplits.addAll(splits);
+    waitingForSplitChangeFuture.complete(null);
```

Comment:
In general, there is no guarantee that `notifyNoMoreSplits` is always going to be invoked, or when it will be invoked. 

I think we need to complete the future in both `addSplits()` and `notifyNoMoreSplits()`. Basically if a reader goes to sleep because it has exhausted all the splits, it needs to be waken up either when a new split is assigned or NoMoreSplits notification is received.

Our current `SplitEnumerator` implementation follows a static splits assignment approach, so it sends all the splits to a subtask at once and then sends the `NoMoreSplits` notification immediately. So timing wise, it seems that the reader can wait for the `NoMoreSplits` notification and then act. However, that won't work for the dynamic assignment case. For example, if a reader only gets one split at a time and will request another split from the `SplitEnumerator` after finishing reading from the current split, it has to wake up and poll once a split is assigned. So the `splitChangeFuture` has to be completed in `addSplits()`. Also, if the reader has exhausted all the splits and gone to sleep, it has to be waken up upon receiving `NoMoreSplits` notification so it can exit normally.

Orthogonally, it seems the code does have a bug here. If there are live readers, the future returned by `isAvailableForAliveReaders()` should also be completed when there is a split change, either from `addSplits()` or `notifyNoMoreSplits()`. I'll update the patch to fix that.